presto-main/src/main/java/com/facebook/presto/operator/aggregation/OptimizedTypedSet.java

```diff
@@ -0,0 +1,347 @@
+            union(block);
+            return;
+        }
+
+        currentBlockIndex++;
+        ensureBlocksCapacity(currentBlockIndex + 1);
+        blocks[currentBlockIndex] = block;
+
+        int[] positions = new int[positionCount];
+
+        // We need to create a new hash table because the elements in the set need be removed
+        long[] newBlockPositionByHash = initializeHashTable();
+
+        int positionsIndex = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int hash = getMaskedHash(hashPosition(elementType, block, i));
+            int positionInBlockPositionByHash = getInsertPosition(blockPositionByHash, hash, block, i);
+
+            // add to the hash table if it does not exist in blockPositionByHash
+            if (positionInBlockPositionByHash != INVALID_POSITION) {
+                if (addElement(newBlockPositionByHash, hash, block, i)) {
+                    positions[positionsIndex++] = i;
+                }
+            }
+        }
+
+        blockPositionByHash = newBlockPositionByHash;
+        getPositionsForBlocks().add(positionsList(positions, 0, positionsIndex));
+        size = positionsIndex;
+
+        clearPreviousBlocks();
+    }
+
+    /**
+     * Build and return the block representing this set
+     */
+    public Block getBlock()
+    {
+        if (size == 0) {
+            return elementType.createBlockBuilder(null, 0).build();
+        }
+
+        if (currentBlockIndex == 0) {
+            // Just one block. Return a DictionaryBlock
+            Block block = blocks[currentBlockIndex];
+            SelectedPositions selectedPositions = getPositionsForBlocks().get(currentBlockIndex);
+
+            if (!selectedPositions.isList()) {
+                if (selectedPositions.size() == block.getPositionCount()) {
+                    return block;
+                }
+                else {
+                    return block.getRegion(selectedPositions.getOffset(), selectedPositions.size());
+                }
+            }
+            return new DictionaryBlock(
+                    selectedPositions.getOffset(),
+                    selectedPositions.size(),
+                    block,
+                    selectedPositions.getPositions(),
+                    false,
+                    DictionaryId.randomDictionaryId());
+        }
+
+        Block firstBlock = blocks[0];
+        BlockBuilder blockBuilder = elementType.createBlockBuilder(
+                null,
+                size,
+                toIntExact(firstBlock.getApproximateRegionLogicalSizeInBytes(0, firstBlock.getPositionCount()) / max(1, toIntExact(firstBlock.getPositionCount()))));
+        for (int i = 0; i <= currentBlockIndex; i++) {
+            Block block = blocks[i];
+            SelectedPositions selectedPositions = getPositionsForBlocks().get(i);
+            int positionCount = selectedPositions.size();
+
+            if (!selectedPositions.isList()) {
```

Comment:
curious: when would this happen? (in that case we will return the current block positions.

Reply:
> curious: when would this happen? (in that case we will return the current block positions.

@wenleix Nice catch. Right now there isn't. I put the code there in case there is cases where the whole block(without duplicates) is added. Do you think I should remove this?

Reply:
> curious: when would this happen? (in that case we will return the current block positions.

@wenleix I removed the `if (!selectedPositions.isList()) {} ` code block. Thank you for the catch. Will you be able to take another look? Thank you!